receiver/vmmetricsreceiver/vm_metrics_collector.go

```diff
@@ -113,104 +106,137 @@ func (vmc *VMMetricsCollector) StopCollection() {
+	defer span.End()
+
+	metrics := make([]*metricspb.Metric, 0, len(vmMetricDescriptors))
+	var errs []error
+
 	ms := &runtime.MemStats{}
 	runtime.ReadMemStats(ms)
-	ctx := context.Background()
-	stats.Record(
-		ctx,
-		mRuntimeAllocMem.M(int64(ms.Alloc)),
-		mRuntimeTotalAllocMem.M(int64(ms.TotalAlloc)),
-		mRuntimeSysMem.M(int64(ms.Sys)))
+	metrics = append(
+		metrics,
+		&metricspb.Metric{
+			MetricDescriptor: metricAllocMem,
+			Timeseries:       []*metricspb.TimeSeries{vmc.getInt64TimeSeries(ms.Alloc)},
+		},
+		&metricspb.Metric{
+			MetricDescriptor: metricTotalAllocMem,
+			Timeseries:       []*metricspb.TimeSeries{vmc.getInt64TimeSeries(ms.TotalAlloc)},
+		},
+		&metricspb.Metric{
+			MetricDescriptor: metricSysMem,
+			Timeseries:       []*metricspb.TimeSeries{vmc.getInt64TimeSeries(ms.Sys)},
+		},
+	)
 
 	pid := os.Getpid()
 	var proc procfs.Proc
 	var err error
-	if vmc.processFs == "" {
-		proc, err = vmc.fs.NewProc(pid)
-	} else {
-		proc, err = vmc.processFs.NewProc(pid)
-	}
+	proc, err = vmc.processFs.NewProc(pid)
 	if err == nil {
 		procStat, err := proc.NewStat()
 		if err == nil {
+			var processCPUTimeDelta float64
 			if prevProcStat != nil {
-				stats.Record(ctx, mCPUSeconds.M(procStat.CPUTime()-prevProcStat.CPUTime()))
+				processCPUTimeDelta = procStat.CPUTime() - prevProcStat.CPUTime()
 			} else {
-				stats.Record(ctx, mCPUSeconds.M(procStat.CPUTime()))
+				processCPUTimeDelta = procStat.CPUTime()
 			}
+			metrics = append(
+				metrics,
+				&metricspb.Metric{
+					MetricDescriptor: metricProcessCPUSeconds,
+					Timeseries:       []*metricspb.TimeSeries{vmc.getDoubleTimeSeries(processCPUTimeDelta, nil)},
+				},
+			)
 		}
 		prevProcStat = &procStat
 	}
 
+	if err != nil {
+		errs = append(errs, err)
+	}
+
 	stat, err := vmc.fs.NewStat()
 	if err == nil {
 		cpuStat := stat.CPUTotal
-		stats.Record(
-			ctx,
-			mProcessesRunning.M(int64(stat.ProcessesRunning)),
-			mProcessesBlocked.M(int64(stat.ProcessesBlocked)))
+		var processCreatedDelta uint64
+		var cpuTimeUserDelta, cpuTimeSystemDelta, cpuTimeIdleDelta, cpuTimeNiceDelta, cpuTimeIOWaitDelta float64
 		if prevStat != nil {
-			stats.Record(
-				ctx,
-				mProcessesCreated.M(int64(stat.ProcessCreated-prevStat.ProcessCreated)),
-				mUserCPUSeconds.M(cpuStat.User-prevStat.CPUTotal.User),
-				mNiceCPUSeconds.M(cpuStat.Nice-prevStat.CPUTotal.Nice),
-				mSystemCPUSeconds.M(cpuStat.System-prevStat.CPUTotal.System),
-				mIdleCPUSeconds.M(cpuStat.Idle-prevStat.CPUTotal.Idle),
-				mIowaitCPUSeconds.M(cpuStat.Iowait-prevStat.CPUTotal.Iowait))
+			processCreatedDelta = stat.ProcessCreated - prevStat.ProcessCreated
+			cpuTimeUserDelta = cpuStat.User - prevStat.CPUTotal.User
+			cpuTimeSystemDelta = cpuStat.System - prevStat.CPUTotal.System
+			cpuTimeIdleDelta = cpuStat.Idle - prevStat.CPUTotal.Idle
+			cpuTimeNiceDelta = cpuStat.Nice - prevStat.CPUTotal.Nice
+			cpuTimeIOWaitDelta = cpuStat.Iowait - prevStat.CPUTotal.Iowait
```

Comment:
don't need delta if it is directly exported.

Reply:
You're right - good catch.